framework/src/userobject/DomainUserObject.C

```diff
@@ -64,49 +64,64 @@ DomainUserObject::DomainUserObject(const InputParameters & parameters)
 
     for (const auto interface_bnd_id : _interface_bnd_ids)
     {
+      bool has_connection = false;
       const auto & interface_connected_blocks = _mesh.getInterfaceConnectedBlocks(interface_bnd_id);
       for (const auto interface_connected_block : interface_connected_blocks)
       {
         if (hasBlocks(interface_connected_block))
-          // we're operating on this block
-          continue;
-
-        // these are blocks connected to our blocks
-        _interface_connected_blocks.insert(interface_connected_block);
+          has_connection = true;
+        else
+          // these are blocks connected to our blocks
+          _interface_connected_blocks[interface_bnd_id].insert(interface_connected_block);
       }
+      if (!has_connection)
+        paramError("interface_boundaries",
+                   "Interface with ID ",
+                   interface_bnd_id,
+                   " is not connected with the domain of the domain user object ",
+                   name());
     }
   }
 }
 
 const MooseVariableFieldBase *
-DomainUserObject::getInterfaceFieldVar(const std::string & var_name, const unsigned int comp)
+DomainUserObject::getInterfaceFieldVar(const std::string & var_name,
+                                       const unsigned int comp,
+                                       const std::set<BoundaryID> * interfaces)
 {
   const auto * const field_var = getFieldVar(var_name, comp);
   mooseAssert(field_var, "We should not be able to return a null variable");
-  _interface_vars.insert(field_var);
+  if (interfaces)
+    _var_interfaces[field_var->name()] = *interfaces;
+  else
+    _var_interfaces[field_var->name()] = _interface_bnd_ids;
   return field_var;
 }
 
 void
 DomainUserObject::checkVariable(const MooseVariableFieldBase & variable) const
 {
-  if (_interface_vars.count(&variable))
+  auto it = _var_interfaces.find(variable.name());
+  if (it != _var_interfaces.end())
   {
     // we could have done this check in the constructor but to be consistent with other block
     // restrictable checks, we'll do it here
-    for (const auto connected_block : _interface_connected_blocks)
-      if (!variable.hasBlocks(connected_block))
-        mooseError("Variable '",
-                   variable.name(),
-                   "' is not defined on the interface connected block '",
-                   _mesh.getSubdomainName(connected_block),
-                   "'");
+    auto & bnd_ids = it->second;
+    for (const auto bnd_id : bnd_ids)
+    {
+      const auto & connected_blocks = _interface_connected_blocks.at(bnd_id);
+      for (const auto & bid : connected_blocks)
+        if (!variable.hasBlocks(bid))
+          mooseError("Variable '",
+                     variable.name(),
+                     "' is not defined on the interface connected block '",
+                     _mesh.getSubdomainName(bid),
+                     "'");
```

Comment:
but that s part of the point of interfaces right. an interface UO may be the interface between two different quantities, defined on each side

Reply:
The application developer will specify a parameter like
```c++
  params.addRequiredCoupledVar("v", "The variable on the other side of the interface");
```
and then we retrieve that kind of variable with a `getInterfaceFieldVar` API which marks it as a "special" variable for which we want to check integrity in a different way than the typical `BlockRestrictable` check which is meant for variables defined on the same blocks as the `DomainUserObject`